### PR TITLE
[LOG4J2-2399] support for async formatting on reusable messages

### PR DESCRIPTION
**Pull request by bjlaub**
Adds support for deferring message content formatting on certain
ReusableMessage implementations to a background thread when async
logging is enabled. This change adds the ability for log events to
retain an instance of a formatter for a message so that the message
content can be formatted later, after the message has already been
recycled.

This is an effort to optimize performance for use cases where formatting
on the logging thread is unnecessary or undesired.

TODO:
- [ ] benchmarks

---

Re-opened from #205

### CODE REVIEW

**Comment by remkop** [comment]
Would it be possible to see some JMH benchmarks that compare performance with and without this change?

**Comment by bjlaub** [comment]
@remkop @cakofony if i'm understanding correctly the goal here is to make sure the work of formatting messages is moved off of the application logging thread and on to a background thread. I've tried to add a benchmark to demonstrate that - the total amount of work ends up the same in practice, but the benchmark should show that we're offloading the work of formatting to a background thread when using ReusableMessages and async formatting. Results are below:

```
# Run complete. Total time: 00:06:05

Benchmark                                                                                            (asyncFormatting)             (messageFactory)    Mode     Cnt         Score     Error  Units
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams                                               false  ParameterizedMessageFactory  sample  983443      3963.993 ± 122.528  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.00                false  ParameterizedMessageFactory  sample              3244.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.50                false  ParameterizedMessageFactory  sample              3488.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.90                false  ParameterizedMessageFactory  sample              3724.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.95                false  ParameterizedMessageFactory  sample              4296.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.99                false  ParameterizedMessageFactory  sample             12224.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.999               false  ParameterizedMessageFactory  sample             25728.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.9999              false  ParameterizedMessageFactory  sample             55913.958            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p1.00                false  ParameterizedMessageFactory  sample           7651328.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams                                               false       ReusableMessageFactory  sample  904581      4452.558 ± 553.731  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.00                false       ReusableMessageFactory  sample              3132.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.50                false       ReusableMessageFactory  sample              3500.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.90                false       ReusableMessageFactory  sample              4832.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.95                false       ReusableMessageFactory  sample              5416.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.99                false       ReusableMessageFactory  sample             16640.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.999               false       ReusableMessageFactory  sample             41216.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.9999              false       ReusableMessageFactory  sample            176149.402            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p1.00                false       ReusableMessageFactory  sample          85590016.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams                                                true  ParameterizedMessageFactory  sample  813313      1322.547 ± 166.508  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.00                 true  ParameterizedMessageFactory  sample               935.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.50                 true  ParameterizedMessageFactory  sample              1038.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.90                 true  ParameterizedMessageFactory  sample              1264.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.95                 true  ParameterizedMessageFactory  sample              1538.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.99                 true  ParameterizedMessageFactory  sample              4192.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.999                true  ParameterizedMessageFactory  sample             23061.952            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.9999               true  ParameterizedMessageFactory  sample             48597.581            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p1.00                 true  ParameterizedMessageFactory  sample          21561344.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams                                                true       ReusableMessageFactory  sample  612640       449.432 ±   4.292  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.00                 true       ReusableMessageFactory  sample               273.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.50                 true       ReusableMessageFactory  sample               381.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.90                 true       ReusableMessageFactory  sample               498.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.95                 true       ReusableMessageFactory  sample               549.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.99                 true       ReusableMessageFactory  sample               791.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.999                true       ReusableMessageFactory  sample             13621.744            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p0.9999               true       ReusableMessageFactory  sample             27040.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8ComplexParams:throughput8ComplexParams·p1.00                 true       ReusableMessageFactory  sample            384512.000            ns/op
```

**Comment by remkop** [comment]
So, to put it in words, it looks like async formatting has about 3x better latency with `ParameterizedMessageFactory` and about 10x better latency with `ReusableMessageFactory`.
Is that a good summary of these numbers?

**Comment by cakofony** [comment]
That is correct for this input, though it is entirely dependent on the cost of converting parameters into string values.

A benchmark with eight string values rather than beans with allocation in the tostring implementation is much closer, though both of the asynchronously formatted benchmarks are identical, because the work on the application thread is the same:

```
Benchmark                                                                                            (asyncFormatting)             (messageFactory)    Mode     Cnt         Score     Error  Units
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams                                              false  ParameterizedMessageFactory  sample  227620      2754.177 ± 411.508  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.00                false  ParameterizedMessageFactory  sample              2036.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.50                false  ParameterizedMessageFactory  sample              2252.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.90                false  ParameterizedMessageFactory  sample              2492.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.95                false  ParameterizedMessageFactory  sample              2832.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.99                false  ParameterizedMessageFactory  sample              5792.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.999               false  ParameterizedMessageFactory  sample             23968.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.9999              false  ParameterizedMessageFactory  sample             51095.994            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p1.00                false  ParameterizedMessageFactory  sample          16596992.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams                                              false       ReusableMessageFactory  sample  230147      2444.984 ±  15.601  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.00                false       ReusableMessageFactory  sample              1938.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.50                false       ReusableMessageFactory  sample              2264.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.90                false       ReusableMessageFactory  sample              2456.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.95                false       ReusableMessageFactory  sample              2580.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.99                false       ReusableMessageFactory  sample              5008.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.999               false       ReusableMessageFactory  sample             23328.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.9999              false       ReusableMessageFactory  sample             53684.634            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p1.00                false       ReusableMessageFactory  sample            702464.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams                                               true  ParameterizedMessageFactory  sample  247582      1183.063 ±   8.424  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.00                 true  ParameterizedMessageFactory  sample               941.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.50                 true  ParameterizedMessageFactory  sample              1036.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.90                 true  ParameterizedMessageFactory  sample              1130.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.95                 true  ParameterizedMessageFactory  sample              1212.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.99                 true  ParameterizedMessageFactory  sample              3512.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.999                true  ParameterizedMessageFactory  sample             20941.344            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.9999               true  ParameterizedMessageFactory  sample             34140.813            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p1.00                 true  ParameterizedMessageFactory  sample            108544.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams                                               true       ReusableMessageFactory  sample  195230       441.564 ±   8.628  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.00                 true       ReusableMessageFactory  sample               270.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.50                 true       ReusableMessageFactory  sample               366.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.90                 true       ReusableMessageFactory  sample               432.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.95                 true       ReusableMessageFactory  sample               492.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.99                 true       ReusableMessageFactory  sample              1258.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.999                true       ReusableMessageFactory  sample             18848.000            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p0.9999               true       ReusableMessageFactory  sample             30772.650            ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput8StringParams:throughput8StringParams·p1.00                 true       ReusableMessageFactory  sample            242944.000            ns/op
```

A test with fewer parameters shows a much smaller difference:

Updated to `log("Hello {}", "World")`

```
Benchmark                                                                                        (asyncFormatting)             (messageFactory)    Mode     Cnt       Score    Error  Units
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam                                             false  ParameterizedMessageFactory  sample  313489     535.499 ±  7.370  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.00                false  ParameterizedMessageFactory  sample             368.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.50                false  ParameterizedMessageFactory  sample             454.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.90                false  ParameterizedMessageFactory  sample             520.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.95                false  ParameterizedMessageFactory  sample             596.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.99                false  ParameterizedMessageFactory  sample            1124.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.999               false  ParameterizedMessageFactory  sample           19104.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.9999              false  ParameterizedMessageFactory  sample           53600.640           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p1.00                false  ParameterizedMessageFactory  sample          130432.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam                                             false       ReusableMessageFactory  sample  323033     509.602 ± 10.065  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.00                false       ReusableMessageFactory  sample             290.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.50                false       ReusableMessageFactory  sample             426.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.90                false       ReusableMessageFactory  sample             575.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.95                false       ReusableMessageFactory  sample             630.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.99                false       ReusableMessageFactory  sample            1068.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.999               false       ReusableMessageFactory  sample           16800.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.9999              false       ReusableMessageFactory  sample           38904.077           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p1.00                false       ReusableMessageFactory  sample          619520.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam                                              true  ParameterizedMessageFactory  sample  192159     463.767 ±  8.593  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.00                 true  ParameterizedMessageFactory  sample             255.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.50                 true  ParameterizedMessageFactory  sample             386.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.90                 true  ParameterizedMessageFactory  sample             452.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.95                 true  ParameterizedMessageFactory  sample             516.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.99                 true  ParameterizedMessageFactory  sample            1476.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.999                true  ParameterizedMessageFactory  sample           17109.760           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.9999               true  ParameterizedMessageFactory  sample           43423.232           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p1.00                 true  ParameterizedMessageFactory  sample          131584.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam                                              true       ReusableMessageFactory  sample  221207     381.790 ±  5.879  ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.00                 true       ReusableMessageFactory  sample             180.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.50                 true       ReusableMessageFactory  sample             320.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.90                 true       ReusableMessageFactory  sample             419.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.95                 true       ReusableMessageFactory  sample             472.000           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.99                 true       ReusableMessageFactory  sample            1119.840           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.999                true       ReusableMessageFactory  sample           14268.032           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p0.9999               true       ReusableMessageFactory  sample           29352.806           ns/op
BackgroundFormattingAsyncLoggersBenchmark.throughput1StringParam:throughput1StringParam·p1.00                 true       ReusableMessageFactory  sample          123648.000           ns/op
```

Is the additional complexity worthwhile? I defer to your judgement @remkop.

**Comment by rgoers** [comment]
@remkop Do you have an opinion on this? It has been almost 2 years now.

File: log4j-api/src/main/java/org/apache/logging/log4j/message/ReusableParameterizedMessage.java [hunk 1]
```diff
@@ -355,4 +356,41 @@ public void clear() { // LOG4J2-1583
...
+    @Override
+    public MessageContentFormatter getMessageContentFormatter() {
+        return Formatter.INSTANCE;
+    }
+
+    private enum Formatter implements MessageContentFormatter {
+        INSTANCE;
+
+        private static final ThreadLocal<int[]> localIndices = new ThreadLocal<int[]>() {
+            @Override
```
Comment (remkop) [inline log4j-api/src/main/java/org/apache/logging/log4j/message/ReusableParameterizedMessage.java:369]: This pattern (subclassing ThreadLocal to provide an initial value) is problematic for frameworks that use class loaders. If  I remember correctly, Tomcat reports this as a potential memory leak. The reason is that the ThreadLocal map now has a reference to a Log4j class (the anonymous ThreadLocal subclass) that is never removed, preventing the garbage collector from GC-ing a web app that was stopped or restarted.

A better way is to simply declare the threadlocal as

```java
private static final ThreadLocal<int[]> localIndices = new ThreadLocal<int[]>();
```
... and providing an accessor (getter) method where the initial value is set. The code using this threadlocal must always use the accessor. This is also not great, but prevents the memory leak.

Comment (rgoers) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/ReusableParameterizedMessage.java:369]: Thanks. @bjlaub if you are still interested is it possible you could make the recommended changes?

**Comment by remkop** [comment]
I did not have time to do an in-depth review today, would need to take another detailed look.

It has been a while since I looked, perhaps @bjlaub  can clarify whether my understanding of the points below is correct?
* The goal of this PR is to improve performance for scenarios where garbage-free logging is combined with Async Loggers. This is achieved by doing some formatting work, which is usually done in the application thread, in the background thread instead. It should _only_ impact cases where garbage-free logging is combined with Async Loggers, is that correct? - Note that I need to take a more detailed look to see how the PR actually works.
* This logic only activates if `Constants.FORMAT_MESSAGES_IN_BACKGROUND` (system property `log4j.format.msg.async`) is `true`.  Is that correct?

Overall:

Note that using this FORMAT_MESSAGES_IN_BACKGROUND configuration is a fairly risky thing to do for applications: the application developers need to be _very_ sure that _none_ of the objects given as parameters to logging calls are mutated after the logging call, or the log may contain a different value from what the actual value was when the logging call was made. If someone who is not aware of this configuration is troubleshooting some problem, then this can be highly confusing. I admit I am responsible for this :sweat_smile:  since I implemented an early version of this configuration but I am no longer sure that this was a good idea... 

Another, related, consideration is that this area (async loggers and garbage-free logging) is complex and a bit fragile. Again, this is my responsibility and not criticism of @bjlaub or anyone else. However, because of this, I believe it is easy to introduce bugs, and so perhaps the bar for accepting changes in this area needs to be a bit higher than usual. I am not yet sure that even a 10x throughput improvement in the formatting is a good trade-off when this advantage can only be achieved by the rare applications that are willing to switch on this (risky) FORMAT_MESSAGES_IN_BACKGROUND configuration.

**Comment by garydgregory** [comment]
On Wed, Jul 29, 2020 at 5:02 AM Remko Popma <notifications@github.com>
wrote:

> I did not have time to do an in-depth review today, would need to take
> another detailed look.
>
> It has been a while since I looked, perhaps @bjlaub
> <https://github.com/bjlaub> can clarify whether my understanding of the
> points below is correct?
>
>    - The goal of this PR is to improve performance for scenarios where
>    garbage-free logging is combined with Async Loggers. This is achieved by
>    doing some formatting work, which is usually done in the application
>    thread, in the background thread instead. It should *only* impact
>    cases where garbage-free logging is combined with Async Loggers, is that
>    correct? - Note that I need to take a more detailed look to see how the PR
>    actually works.
>    - This logic only activates if Constants.FORMAT_MESSAGES_IN_BACKGROUND
>    (system property log4j.format.msg.async) is true. Is that correct?
>
> Overall:
>
> Note that using this FORMAT_MESSAGES_IN_BACKGROUND configuration is a
> fairly risky thing to do for applications: the application developers need
> to be *very* sure that *none* of the objects given as parameters to
> logging calls are mutated after the logging call, or the log may contain a
> different value from what the actual value was when the logging call was
> made. If someone who is not aware of this configuration is troubleshooting
> some problem, then this can be highly confusing. I admit I am responsible
> for this 😅 since I implemented an early version of this configuration
> but I am no longer sure that this was a good idea...
>
> Another, related, consideration is that this area (async loggers and
> garbage-free logging) is complex and a bit fragile. Again, this is my
> responsibility and not criticism of @bjlaub <https://github.com/bjlaub>
> or anyone else. However, because of this, I believe it is easy to introduce
> bugs, and so perhaps the bar for accepting changes in this area needs to be
> a bit higher than usual. I am not yet sure that even a 10x throughput
> improvement in the formatting is a good trade-off when this advantage can
> only be achieved by the rare applications that are willing to switch on
> this (risky) FORMAT_MESSAGES_IN_BACKGROUND configuration.
>

This feature seems like a land mine waiting for activation in any
non-trivial software stack, especially as activated by a global setting any
code and toggle. I would be OK with it being a feature I enabled in a
LoggerContext or better yet in a MessageFactory.

Gary

**Comment by rgoers** [comment]
Does anyone have an interest in doing something with this?

**Comment by ppkarwasz** [comment]
This was closed automatically by Github because we renamed the `release-2.x` branch to `2.x`. Feel free to resubmit to the `2.x` branch.

**Pull request closed without merging**